### ISSUE TITLE
[Bug]: clang crash

### ISSUE THREAD

**Issue by MaxKellermann** [android/ndk#1716]
### Description

Source code: `extern "C" char sinf();int main() { return sinf(); }`
Command line: `~/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -c -target armv7-linux-androideabi test.cpp`
Crashes:
```
PLEASE submit a bug report to https://github.com/android-ndk/ndk/issues and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ -target armv7-linux-androideabi test.cpp -c
1.	<eof> parser at end of file
2.	test.cpp:1:28: LLVM IR generation of declaration 'main'
3.	test.cpp:1:28: Generating code for declaration 'main'
 #0 0x000000000456d858 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x456d858)
 #1 0x000000000456d6e0 llvm::sys::RunSignalHandlers() (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x456d6e0)
 #2 0x000000000453d363 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x453d363)
 #3 0x000000000453d6c1 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x453d6c1)
 #4 0x00007f3c54638200 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12200)
 #5 0x00000000033451a2 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x33451a2)
 #6 0x00000000033977c0 clang::CodeGen::CodeGenFunction::EmitScalarExpr(clang::Expr const*, bool) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x33977c0)
 #7 0x00000000033aca89 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x33aca89)
 #8 0x000000000338e00a clang::CodeGen::CodeGenFunction::EmitBuiltinExpr(clang::GlobalDecl, unsigned int, clang::CallExpr const*, clang::CodeGen::ReturnValueSlot) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x338e00a)
 #9 0x0000000003348569 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3348569)
#10 0x0000000003343010 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3343010)
#11 0x00000000034fb0e0 clang::CodeGen::CodeGenFunction::EmitReturnStmt(clang::ReturnStmt const&) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x34fb0e0)
#12 0x00000000034f578e clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x34f578e)
#13 0x00000000034f563f clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x34f563f)
#14 0x00000000034f4d1c clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x34f4d1c)
#15 0x00000000034f34b0 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x34f34b0)
#16 0x00000000034f2cba clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x34f2cba)
#17 0x0000000002a148ed clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x2a148ed)
#18 0x0000000004875d8f (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x4875d8f)
#19 0x0000000004874e6c (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x4874e6c)
#20 0x000000000389e01a clang::ParseAST(clang::Sema&, bool, bool) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x389e01a)
#21 0x00000000039383d1 clang::FrontendAction::Execute() (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x39383d1)
#22 0x0000000003937fcb clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3937fcb)
#23 0x0000000003937bc1 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3937bc1)
#24 0x0000000003936bff cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3936bff)
#25 0x0000000003e8b9b7 (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3e8b9b7)
#26 0x0000000005912b02 clang::driver::CC1Command::setEnvironment(llvm::ArrayRef<char const*>) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x5912b02)
#27 0x0000000003c35e67 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3c35e67)
#28 0x0000000003c35c4d clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*, bool*) const (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3c35c4d)
#29 0x0000000003acc724 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x3acc724)
#30 0x000000000383e033 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::__1::pair<int, clang::driver::Command const*> >&) (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x383e033)
#31 0x000000000383a5b7 main (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x383a5b7)
#32 0x00007f3c541e67fd __libc_start_main ./csu/../csu/libc-start.c:332:16
#33 0x000000000582fa99 _start (/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang+++0x582fa99)
clang-14: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Android (8075178, based on r437112b) clang version 14.0.1 (https://android.googlesource.com/toolchain/llvm-project 8671348b81b95fc603505dfc881b45103bee1731)
Target: armv7-unknown-linux-android
Thread model: posix
InstalledDir: /home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin
clang-14: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-14: note: diagnostic msg: /tmp/test-07fb0d.cpp
clang-14: note: diagnostic msg: /tmp/test-07fb0d.sh
clang-14: note: diagnostic msg: 

********************
```
`/tmp/test-07fb0d.cpp`:
```
# 1 "<built-in>"
# 1 "test.cpp"
extern "C" char sinf();int main() { return sinf(); }
```
`/tmp/test-07fb0d.sh`:
```
# Crash reproducer for Android (8075178, based on r437112b) clang version 14.0.1 (https://android.googlesource.com/toolchain/llvm-project 8671348b81b95fc603505dfc881b45103bee1731)
# Driver args: "--driver-mode=g++" "--target=armv7-linux-androideabi" "test.cpp" "-c"
# Original command:  "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang-14" "-cc1" "-triple" "armv7-unknown-linux-android" "-emit-obj" "-mrelax-all" "--mrelax-relocations" "-mnoexecstack" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test.cpp" "-mrelocation-model" "pic" "-pic-level" "1" "-fhalf-no-semantic-interposition" "-mframe-pointer=all" "-fno-rounding-math" "-mconstructor-aliases" "-target-cpu" "generic" "-target-feature" "+soft-float-abi" "-target-feature" "+vfp2" "-target-feature" "+vfp2sp" "-target-feature" "+vfp3" "-target-feature" "+vfp3d16" "-target-feature" "+vfp3d16sp" "-target-feature" "+vfp3sp" "-target-feature" "-fp16" "-target-feature" "-vfp4" "-target-feature" "-vfp4d16" "-target-feature" "-vfp4d16sp" "-target-feature" "-vfp4sp" "-target-feature" "-fp-armv8" "-target-feature" "-fp-armv8d16" "-target-feature" "-fp-armv8d16sp" "-target-feature" "-fp-armv8sp" "-target-feature" "-fullfp16" "-target-feature" "+fp64" "-target-feature" "+d32" "-target-feature" "+neon" "-target-feature" "-sha2" "-target-feature" "-aes" "-target-feature" "-fp16fml" "-target-abi" "aapcs-linux" "-mfloat-abi" "soft" "-fallow-half-arguments-and-returns" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/tmp" "-resource-dir" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/14.0.1" "-internal-isystem" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/include/c++/v1" "-internal-isystem" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/lib64/clang/14.0.1/include" "-internal-isystem" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/local/include" "-internal-externc-isystem" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/include/arm-linux-androideabi" "-internal-externc-isystem" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/include" "-internal-externc-isystem" "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/include" "-fdeprecated-macro" "-fdebug-compilation-dir=/tmp" "-ferror-limit" "19" "-fno-signed-char" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-o" "test.o" "-x" "c++" "test.cpp"
 "/home/max/opt/android-ndk-r24/toolchains/llvm/prebuilt/linux-x86_64/bin/clang-14" "-cc1" "-triple" "armv7-unknown-linux-android" "-emit-obj" "-mrelax-all" "--mrelax-relocations" "-mnoexecstack" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test.cpp" "-mrelocation-model" "pic" "-pic-level" "1" "-fhalf-no-semantic-interposition" "-mframe-pointer=all" "-fno-rounding-math" "-mconstructor-aliases" "-target-cpu" "generic" "-target-feature" "+soft-float-abi" "-target-feature" "+vfp2" "-target-feature" "+vfp2sp" "-target-feature" "+vfp3" "-target-feature" "+vfp3d16" "-target-feature" "+vfp3d16sp" "-target-feature" "+vfp3sp" "-target-feature" "-fp16" "-target-feature" "-vfp4" "-target-feature" "-vfp4d16" "-target-feature" "-vfp4d16sp" "-target-feature" "-vfp4sp" "-target-feature" "-fp-armv8" "-target-feature" "-fp-armv8d16" "-target-feature" "-fp-armv8d16sp" "-target-feature" "-fp-armv8sp" "-target-feature" "-fullfp16" "-target-feature" "+fp64" "-target-feature" "+d32" "-target-feature" "+neon" "-target-feature" "-sha2" "-target-feature" "-aes" "-target-feature" "-fp16fml" "-target-abi" "aapcs-linux" "-mfloat-abi" "soft" "-fallow-half-arguments-and-returns" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/tmp" "-fdeprecated-macro" "-fdebug-compilation-dir=/tmp" "-ferror-limit" "19" "-fno-signed-char" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "c++" "test-07fb0d.cpp"
```
(GitHub rejects my upoad "We don't support that file type.")

### Affected versions

r24

### Host OS

Linux

### Host OS version

Debian Bookworm

### Affected ABIs

armeabi-v7a, arm64-v8a, x86, x86_64

**Comment by DanAlbert** [comment]
> (GitHub rejects my upoad "We don't support that file type.")

Is see you attached the inputs so don't need to this time, but for the future, GitHub will accept a zip file. You can also just rename files with `.txt` extensions and it'll allow those too. No need to do that this time since we have the info we need, but if in the future you have large test inputs that's the way to go :)

**Comment by jfgoog** [comment]
The crash also does not occur with 25.0.8528842.

`clang -v` reports:

```
/usr/local/google/home/jamesfarrell/Android/Sdk/ndk/25.0.8355429/toolchains/llvm/prebuilt/linux-x86_64/bin/clang-14 -v
Android (8075178, based on r437112b) clang version 14.0.1 (https://android.googlesource.com/toolchain/llvm-project 8671348b81b95fc603505dfc881b45103bee1731)

/usr/local/google/home/jamesfarrell/Android/Sdk/ndk/25.0.8528842/toolchains/llvm/prebuilt/linux-x86_64/bin/clang-14 -v
Android (8490178, based on r450784d) clang version 14.0.6 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0cca074e9238af8b4106c30add4418f6)
```

**Comment by jfgoog** [comment]
Weirdly, when I sync toolchain/llvm-project to 8671348b81b95fc603505dfc881b45103bee1731 and build clang, I don't get a crash.

**Comment by stephenhines** [comment]
@jfgoog https://android.googlesource.com/platform/prebuilts/clang/host/linux-x86/+/4365d4dfb540abaeb382aedd5a3231e21ecd5b09/clang-r437112b/clang_source_info.md is the list of patches that were applied to the LLVM sources there. It's possible that one of them causes this particular problem, especially since this seems like the kind of crash that really shouldn't happen (for simple code).

**Comment by stephenhines** [comment]
I trimmed the arguments down to:

```
"-cc1" "-triple" "armv7-unknown-linux-android" "-emit-obj" "-discard-value-names" "-main-file-name" "test.cpp" "-target-cpu" "generic" "-target-feature" "+soft-float-abi" "-target-feature" "+vfp2" "-target-feature" "+vfp2sp" "-target-feature" "+vfp3" "-target-feature" "+vfp3d16" "-target-feature" "+vfp3d16sp" "-target-feature" "+vfp3sp" "-target-feature" "-fp16" "-target-feature" "-vfp4" "-target-feature" "-vfp4d16" "-target-feature" "-vfp4d16sp" "-target-feature" "-vfp4sp" "-target-feature" "-fp-armv8" "-target-feature" "-fp-armv8d16" "-target-feature" "-fp-armv8d16sp" "-target-feature" "-fp-armv8sp" "-target-feature" "+fp64" "-target-feature" "+d32" "-target-feature" "+neon" "-target-abi" "aapcs-linux" "-mfloat-abi" "soft" "-fallow-half-arguments-and-returns" "-x" "c++"
```

**Comment by jfgoog** [comment]
Tried a few patches that I missed. Those don't crash either.

`for foo in ../llvm_android/patches/cherry/8901f8beea3a70f92be8c0b80313260502f03727.patch ../llvm_android/patches/cherry/d7e089f2d6a5cd5f283a90ab29241d20d4fc3ed1.patch ../llvm_android/patches/cherry/7dbbb5d3a46e1526cfa126ae02a5856d7ce0fda9.patch ../llvm_android/patches/cherry/da66263b6e505a4c605efbe8d68c3b09ad3606a4.patch ../llvm_android/patches/cherry/dc9f65be4555406262ff693c8bac5f1f0b960a97.patch ../llvm_android/patches/cherry/cce4a7258b81159e57a411896011ee2742f17def.patch ; do echo $foo ; git restore . ; git apply $foo ; ninja -C build clang ; ~/bugs/1716-sinf-crash/bisect.sh ; git restore . ; done`

**Comment by jfgoog** [comment]
OK, I built the compiler properly, using the instructions at https://android.googlesource.com/toolchain/llvm_android/+/master/README.md.

I built it at revision 8075178, which is the same version reported by both version 25.0.8355429 and 24.0.8215888 of the NDK.

I tried the all the patches one at a time, and identified https://source.corp.google.com/android/toolchain/llvm_android/patches/Revert-Sema-Introduce-BuiltinAttr-per-declaration-bu.v4.patch as the culprit.

Without this patch, the test case compiles and does not crash.
